feat(blog): question answering with spokestack and huggingface (#135)
spokestack/spokestack-website: content/blog/building-a-question-answering-bot-with-python/index.md

---
title: Question Answering With Spokestack and Transformers
date: '2021-01-13'
description: This tutorial leverages HuggingFace's Transformer library to make a voice question answering bot with Spokestack.
author: will
tags: Engineering, Tutorial
draft: false
---

> The ability to find information is a fundamental feature of the internet. Often, the goal for this information is to find the answer to a question. When it comes to answering a question about a specific entity, Wikipedia is a useful, accessible, resource for answers. This tutorial is an introduction to how you can use Spokestack and [Huggingface's Transformers](https://huggingface.co/transformers/index.html) library to build a voice interface for a question answering service using data from Wikipedia.

Reviewer: ```suggestion
The ability to find information is a fundamental feature of the internet. Often, the information sought is the answer to a question. When it comes to answering a question about a specific entity, Wikipedia is a useful, accessible, resource. This tutorial will teach you how to use Spokestack and [Huggingface's Transformers](https://huggingface.co/transformers/index.html) library to build a voice interface for a question answering service using data from Wikipedia.
```

> The ability to find information is a fundamental feature of the internet. Often, the goal for this information is to find the answer to a question. When it comes to answering a question about a specific entity, Wikipedia is a useful, accessible, resource for answers. This tutorial is an introduction to how you can use Spokestack and [Huggingface's Transformers](https://huggingface.co/transformers/index.html) library to build a voice interface for a question answering service using data from Wikipedia.

Reviewer 2: it would be good to provide a motivating example of what you will accomplish by completing the tutorial.

## Setting up the Project

First, let's make a directory to hold the project.

```bash
git clone https://github.com/spokestack/wikiqa-python
cd wikiqa-python
```

Now let's set up the python virtual environment. We use `pyenv` to manage virtual environments, but any virtual environment will work.

```bash
pyenv install 3.8.6
pyenv virtualenv 3.8.6 wikiqa
pyenv local wikiqa
```

Then the dependencies.

```bash
pip install -r requirements.txt
```

### TensorFlow

> For this tutorial, we are using the full [tensorflow](http://tensorflow.org) package for a little more functionality than is included with the [TFLite Interpreter](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter). If this is the first time you are installing `tensorflow` you may need to visit the link [here](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter) for system-specific installation. For those who already have the ability to install via [pypi](https://pypi.org/) the following will install the library to your environment.

Reviewer: > you may need to visit the link here...

The text before this is talking about installing TF instead of TFLite, but the link for "here" goes to the TFLite instructions.

> For this tutorial, we are using the full [tensorflow](http://tensorflow.org) package for a little more functionality than is included with the [TFLite Interpreter](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter). If this is the first time you are installing `tensorflow` you may need to visit the link [here](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter) for system-specific installation. For those who already have the ability to install via [pypi](https://pypi.org/) the following will install the library to your environment.

Reviewer 2: Also "here" is never the right thing to do for a link. All links should be for relevant text, eg 

```suggestion
For this tutorial, we are using the full [tensorflow](http://tensorflow.org) package for a little more functionality than is included with the [TFLite Interpreter](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter). If this is the first time you are installing `tensorflow` you should [follow the system-specific installation instructions](https://www.tensorflow.org/lite/guide/python#install_just_the_tensorflow_lite_interpreter). For those who already have the ability to install via [pypi](https://pypi.org/) the following will install the library to your environment.
```

```bash
pip install tensorflow
```

### Transformers

In addition to TensorFlow, you will also need to install [HuggingFace's](https://huggingface.co/transformers/index.html) `transformers` library for the question answering model.

```bash
pip install transformers
```

## Speech Pipeline with Profiles

> Profiles are preset configurations for our [Speech Pipeline](/docs/Python/speech-pipeline). For this tutorial, we will use the Spokestack wakeword and ASR profile. This is everything you need to be able to speak to your app. Simple right?

Reviewer: ```suggestion
Profiles are preset configurations for our [Speech Pipeline](/docs/Python/speech-pipeline). For this tutorial, we will use the Spokestack wakeword and ASR profile. This is everything you need to be able to speak to your app.
```

I don't think that last sentence really adds anything, but if you don't want to take it out, add a comma after "simple".

Since you're using API keys here, we usually take a sentence to say that they'll be needed and include a link to create a free account (`/create`) or log in (`/login`), then say that they can find credentials in `/account/settings`.

```python
from spokestack.profile.wakeword_asr import WakewordSpokestackASR

pipeline = WakewordSpokestackASR.create(
    "spokestack_id", "spokestack_secret", model_dir="path_to_tflite_model_dir"
)
pipeline.start()
```

## Natural Language Understanding (NLU)

Natural Language Understanding (NLU) is how we transform what the user says into action. For more explanation on the NLU see our [docs](/docs/Concepts/nlu). The NLU model for this project is already included in the GitHub repository. However, we will briefly discuss the model configuration in the next section.

```python

> from spokestack.nlu.tflite TFLiteNLU

Reviewer: ```suggestion
from spokestack.nlu.tflite import TFLiteNLU
```

Author: this is why `blacken-docs` couldn't format it...

nlu = TFLiteNLU("tflite")
```

### Model Configuration

> For this model, we will use a basic NLU template which includes intents like `greet`, `accept`, `help`. In addition, we need a way to identify an entity in the user utterance to perform a Wikipedia search (more on this later). The name of this intent is `ask.question` and the templates are simple ways to ask a question. For more details on our NLU data format, you can go to the following link[\]. Naturally, these templates could be more complex and cover a wider variety of utterances, but for the purpose of this tutorial these will be enough.

Reviewer: ```suggestion
For this model, we will use a basic NLU template which includes intents like `greet`, `accept`, and `help`. In addition, we need a way to identify an entity in the user utterance to perform a Wikipedia search (more on this later). The name of this intent is `ask.question`, and the utterance templates are simple ways to ask a question. For more details on our NLU data format, see [here](). Naturally, these templates could be more complex and cover a wider variety of utterances, but for the purpose of this tutorial these will be enough.
```

I think we need a bit more of a transition before this paragraph. The source files used to create the model aren't included in the repository anywhere, so introducing this as "configuration" could be a bit confusing—a reader new to the scene might go looking for it as a way to change how their model performs.

You might want to say something like
> We've included a pre-trained model so you can follow along with this tutorial, but if you want to create your own, here's a quick introduction to writing NLU training data. See our documentation for more information on [our data format](LINK) and how to [train your own model](LINK).

```
[slots.entity]
type = "entity"

[utterances]
values = [
    "who is {entity}?",
    "who is a {entity}?",

>     "who is an {entity}?",

Reviewer: "who is an Marie Curie?"

I can't think of an example where the indefinite article pairs with "who" and a named entity; it's typically more along the lines of, "who's a good boy?"

    "what is {entity}?",
    "what is a {entity}?",
    "what is an {entity}?",
    "{entity}"
]
```

## Question Answering (QA)

> For the QA model, we are using a [SQUAD](https://arxiv.org/abs/1606.05250) style span detection method. The problem is framed as “given this question and this context that contains the possible answer, identify the text span that contains the answer”. At this point, you most likely are thinking "where do we get the context from". We will answer that in the next section.

Reviewer: ```suggestion
For the QA model, we are using a [SQUAD](https://arxiv.org/abs/1606.05250)-style span detection method. The problem is framed as, “given this question and this context that contains the possible answer, identify the text span that contains the answer”. At this point, you most likely are thinking, "where do we get the context?". We will answer that in the next section.
```

### Context Retrieval

> Earlier, we mentioned that we wanted to be able to identify entities in the user's utterance for a Wikipedia search. Context retrieval is the reason for this. Our QA model needs a context from which to draw the answer. Therefore, we will follow the assumption that the question is about a specific entity. For example, "What year was Elon Musk born?". From this utterance, we pull out the entity "Elon Musk" and do a search to retrieve his Wikipedia page. Then, we feed the entire Wikipedia page into the model as context for our question.

Reviewer 2: Ugh please pick a different entity than Space Karen

Author: Will do. I'm glad you pointed this out. It's probably good practice to pick relatively static entities. A lot has happened since I first started working on this.

Reviewer: Wasn't going to pick at the content, but fwiw I agree :).

Reviewer: > Spokestack, who is Space Karen?

_searches twitter for answer_

### Setting Up The Model

We are using a pre-trained [DistilBERT](https://huggingface.co/transformers/model_doc/distilbert.html) model, which is fine-tuned on the [SQUAD](https://arxiv.org/abs/1606.05250) dataset. The choice of model is based on the size, similarity of the fine-tuning task to our objective, and availability at the time of writing. The [collection](https://huggingface.co/models) of pre-trained models is always growing, so feel free to try any of the other models that were fine-tuned on a similar task. You will most likely find a model that works better than this basic one. We would love to hear about the one you discovered, or even better, trained!

```python
from transformers import AutoTokenizer, TFAutoModelForQuestionAnswering

tokenizer = AutoTokenizer.from_pretrained("distilbert-base-cased-distilled-squad")
model = TFAutoModelForQuestionAnswering.from_pretrained("distilbert-base-cased-distilled-squad")
```

## Dialogue Manager

> The dialogue manager for this project is a smart-speaker style interaction. A user asks a question, and the bot speaks the answer. The full example can be seen below. The process starts with retrieving the entity (person/place/thing) found in what the user said. Then, we take that entity and look up its page on Wikipedia. Furthermore, we take the full text of the Wikipedia page and feed it, along with the question, to the QA model. The model gives us the location of the answer, which we use, to grab the answer from the full text. Then, we return the answer in the form of a response. This set up is pretty simple and just meant to get you started. Definitely, expand on this dialogue manager and conform it to your needs.

Reviewer: ```suggestion
The dialogue manager for this project is a smart speaker-style interaction. A user asks a question, and the bot speaks the answer. The full example can be seen below. The process starts with retrieving the entity (person/place/thing) found in what the user said. Then, we take that entity and look up its page on Wikipedia. Next, we take the full text of the Wikipedia page and feed it, along with the question, to the QA model. The model gives us the location of the answer, which we use to grab the answer from the full text. Then, we return the answer in the form of a response. This setup is pretty simple and just meant to get you started. Definitely expand on this dialogue manager and conform it to your needs.
```

```python
"""
Simple QA dialogue manager
"""

import tensorflow as tf
from mediawiki import MediaWiki
from spokestack.nlu.result import Result
from spokestack.nlu.tflite import TFLiteNLU
from transformers import AutoTokenizer, TFAutoModelForQuestionAnswering

class DialogueManager:
    """ Simple Question Answering Dialogue Manager """

    def __init__(self, log_path: str, base_model: str) -> None:
        self._wiki = MediaWiki()
        self._entity_recognizer = TFLiteNLU(log_path)
        self._tokenizer = AutoTokenizer.from_pretrained(base_model)
        self._answerer = TFAutoModelForQuestionAnswering.from_pretrained(base_model)

    def __call__(self, utterance: str) -> str:

>         result = self._entity_recognizer(utterance)

Reviewer: An entity recognizer wouldn't return an intent; I think it's best to leave this named `nlu` or something similar.

            # prepare qa model inputs
            inputs = self._tokenizer(
                result.utterance,
                passage,
                return_tensors="tf",
                padding=True,
                truncation=True,
            )
            # compute answer span
            start_scores, end_scores = self._answerer(inputs)
            start, end = tf.argmax(start_scores, -1)[0], tf.argmax(end_scores, -1)[0]
            # prepare the passage ids for slicing
            tokens = self._tokenizer.convert_ids_to_tokens(
                (inputs["input_ids"].numpy()[0])
            )
            # retrieve only the answer from the passage
            answer = self._tokenizer.convert_tokens_to_string(tokens[start : end + 1])
            return answer
        return "I don't have an answer for that"

    @staticmethod
    def greet() -> str:
        return "Hello, Ask me anything"

    @staticmethod
    def exit() -> str:
        return "Goodbye"

    @staticmethod
    def fallback() -> str:
        return (
            "I'm having trouble understanding your request, could you please "
            "repeat it"
        )

    @staticmethod
    def help() -> str:

>         return "Ask a question like, how long is the amazon river?"

Reviewer: Have you verified that your `ask.question` intent covers questions worded like this?

Author: It did when I tested it, but without the template, it's a lucky guess. Questions like this should be in the template. I will flesh out the template.

Author: @noelweichbrodt suggested using a different entity, and I agree, so I think I may rework the templates to be based on more questions like this.

```

OK, now we have our response from the dialogue manager. The next question on your mind is probably "how do we give it a voice?". Check out the following section for how to set up Spokestack's text to speech service.

## Text to Speech (TTS)

> Now, let's give the app a voice. Similar to the profile section, you will need your Spokestack API keys. We off a `TextToSpeechManager` class with accepts a TTS client and an output source. For most of you, the `PyAudioOuput` class should work. It uses the default system speaker.

Reviewer: ```suggestion
Now, let's give the app a voice. Similar to the profile section, you will need your Spokestack API keys. We offer a `TextToSpeechManager` class which requires a TTS client and an output source. In most cases, the `PyAudioOuput` class should work. It uses the default system speaker.
```

```python
manager = TextToSpeechManager(

>         TextToSpeechClient(KEY_ID, KEY_SECRET), PyAudioOutput()

Reviewer: You have the credentials formatted as Python constants here but as plain strings above; the two mentions should be consistent.

)
manager.synthesize("hello, world", "text", "demo-male")
```

Now that we have given our bot a voice, we can put everything together. At this point, you are just one section away from a voice interface that allows you to get answers to simple questions. Let's move on to the complete working example.

## Putting it All Together

> We have all the modules set up, now we need to add the logic that will respond to events in the conversation.

Reviewer: ```suggestion
We have all the modules set up; now we need to add the logic that will respond to events in the conversation.

```

Author: I completely agree that this pause feels longer than a comma. Can we compromise with a period?

Reviewer: Sure—and it's not so much the theoretical pause length (though that's part of it). Having just a comma makes it a comma splice (a type of [run-on sentence](http://writing.umn.edu/sws/quickhelp/grammar/runons.html)).

Author: Oh right. I didn't even think about them being independent clauses. I guess the theoretical pause should have been a clue.

> For this, we use our [Pipeline Events](/docs/Python/speech-pipeline). To set these up, you want to decorate functions with an event decorator. Pipeline Events are simply events that occur while the pipeline is running. Most applications will want an event that does something when speech is recognized. For ours, we want to process the question and play the response. This is defined in `on_recognize`. The functions we are using for this example can be seen below. For more information on the included events take a look [here](/docs/Python/speech-pipeline).

Reviewer: ```suggestion
For this, we use our [Pipeline Events](/docs/Python/speech-pipeline). Pipeline Events are simply events that occur while the pipeline is running. To use them, you decorate functions with an event decorator. Most applications will want an event hander that does something when speech is recognized. For ours, we want to process the question and play the response. This is defined in the `on_recognize` handler. The functions we are using for this example can be seen below. For more information on the included events take a look [here](/docs/Python/speech-pipeline#speech-event-handlers).
```

```python

> @pipeline.event

Reviewer: You probably don't need to include examples of the activate/deactivate events if all they're going to do is print; readers can check out the other events at the link if they're interested.

## Speech Pipeline with Profiles

> Profiles are preset configurations for our [Speech Pipeline](/docs/Python/speech-pipeline). For this tutorial, we will use the Spokestack wakeword and ASR profile. Now, you will need to get your API credentials. If you already have a free account, [login](/login), if you do not, you will need to [create](/create) one. The credentials can be found in your [account settings](/account/settings). This is everything you need to speak to your app.

Reviewer: ```suggestion
Profiles are preset configurations for our [Speech Pipeline](/docs/Python/speech-pipeline). For this tutorial, we will use the Spokestack wakeword and ASR profile. Now, you will need to get your API credentials. If you already have a free account, [log in](/login); if you do not, you will need to [create](/create) one. The credentials can be found in your [account settings](/account/settings). This is everything you need to speak to your app.
```
(semicolon or period, as before)

```python
from spokestack.profile.wakeword_asr import WakewordSpokestackASR

pipeline = WakewordSpokestackASR.create(

>     KEY_ID, KEY_SECRET, model_dir="path_to_tflite_model_dir"

Reviewer: Just noticing this on this readthrough: You might want to include a quick sentence or two about wakeword models (wakeword is processed on-device/this app uses the default "Spokestack" wakeword) to motivate the need for `model_dir` here. Also, just like the credential naming consistency, might as well use `"tflite"` as `model_dir` here like the NLU section does.
